frb_codegen/src/ir/file.rs

```diff
@@ -1,29 +1,123 @@
+thread_local!(static FETCHED_FOR_SHARED_TYPES_INPUT: RefCell<bool> = RefCell::new(false));
+
+// shared types among all regular API blocks in context of function outputs
+thread_local!(static SHARED_TYPES_OUTPUT: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_OUTPUT: RefCell<bool> = RefCell::new(false));
+
+// shared types among all regular API blocks
+thread_local!(static SHARED_TYPES_ALL: RefCell<HashSet<IrType>> = RefCell::new(HashSet::new()));
+thread_local!(static FETCHED_FOR_SHARED_TYPES_ALL: RefCell<bool> = RefCell::new(false));
+
 #[derive(Debug, Clone)]
 #[cfg_attr(feature = "serde", derive(serde::Serialize))]
 pub struct IrFile {
-    pub funcs: Vec<IrFunc>,
+    /// for a regular API block: this field contains APIs defined in this block and
+    /// all methods of types(struct/enum) used ONLY in this block
+    /// ---NOTE: for shared types used in this block, their methods are also included here;
+    /// for a shared block: it ONLY contains methods from ALL shared types.
+    funcs: Vec<IrFunc>,
     pub struct_pool: IrStructPool,
     pub enum_pool: IrEnumPool,
     pub has_executor: bool,
+    pub block_index: BlockIndex,
+    pub shared: bool, // `true` for an auto-generated shared block, otherwise `false`
 }
 
 impl IrFile {
+    pub fn funcs(&self, exclude_shared_method: bool) -> Vec<IrFunc> {
+        if self.shared {
+            // assert each item of funcs are shared
+            for each in &self.funcs {
+                assert!(each.shared);
+            }
+            return self.funcs.clone();
+        }
+        match exclude_shared_method {
+            true => self
+                .funcs
+                .iter()
+                .cloned()
+                .filter(|each| !each.shared)
+                .collect::<Vec<_>>(),
+            false => self.funcs.clone(),
+        }
+    }
+
+    pub fn set_funcs(&mut self, dst_funcs: Vec<IrFunc>) {
+        self.funcs = dst_funcs;
+    }
+
+    pub fn append_funcs(&mut self, dst_funcs: Vec<IrFunc>) {
+        self.funcs.extend(dst_funcs);
+    }
+
+    /// Returns a vector of all shared methods in the file.
+    /// For a regular API block, this method returns all methods of shared types used in this block.
+    /// For a shared block, it returns all methods from all shared types.
+    pub fn get_shared_methods(&self) -> Vec<IrFunc> {
+        if !self.shared {
+            self.funcs
+                .iter()
+                .cloned()
+                .filter(|each| each.shared)
+                .collect::<Vec<_>>()
+        } else {
+            self.funcs.clone()
+        }
+    }
+
+    pub fn new(
+        funcs: Vec<IrFunc>,
+        struct_pool: IrStructPool,
+        enum_pool: IrEnumPool,
+        has_executor: bool,
+        block_index: BlockIndex,
+        all_configs: &[Opts],
+        shared: bool,
+    ) -> Self {
+        let ir_fie = IrFile {
```

Comment:
ir_fie -> ir_file

Reply:
Refined in [fb58892](https://github.com/fzyzcjy/flutter_rust_bridge/pull/1145/commits/fb588924188eb08f904b3e1db567eaaf01fd2470).